parser/openfoodfacts_taxonomy_parser/parser/taxonomy_parser.py

```diff
@@ -324,6 +326,51 @@ def _harvest_entries(self, filename: str, entries_start_line: int) -> Iterator[N
         data.src_position = line_number + 1 - len(data.preceding_lines)
         yield data
 
+    def _get_valid_child_links(self, entry_nodes: list[NodeData], child_links: list[ChildLink]):
+        """Get only the valid child links, i.e. the child links whose the parent_id exists"""
+        node_ids = set([node.id for node in entry_nodes])
+        # we collect nodes where a synonym was used to designate the parent (unnormalised)
+        unnormalised_child_links = [
+            child_link.copy()
+            for child_link in child_links
+            if child_link["parent_id"] not in node_ids
+        ]
+
+        # we need to group them by language code of the parent id
+        lc_child_links_map = collections.defaultdict(list)
+        for child_link in unnormalised_child_links:
+            lc, parent_id = child_link["parent_id"].split(":")
+            child_link["parent_id"] = parent_id
+            lc_child_links_map[lc].append(child_link)
+
+        # we need to check if the parent id exists in the tags_ids_lc
+        missing_child_links = []
+        for lc, lc_child_links in lc_child_links_map.items():
+            # we collect all the tags_ids in a certain language
+            tags_ids = set()
+            for node in entry_nodes:
+                if "tags_ids_" + lc in node.tags:
+                    tags_ids.update(node.tags["tags_ids_" + lc])
+
+            # we check if the parent_id exists in the tags_ids
+            for child_link in lc_child_links:
+                if child_link["parent_id"] not in tags_ids:
+                    missing_child_links.append((lc, child_link))
```

Comment:
Very good implementation, but it seems of a pity that we do not make the most of this computation.

If, instead of a set, tags_ids is a dict mapping tags_ids to their canonical id, it still can be used to check for existence, but we can also retrieve the real parent at the same cost.

This will simplify (and improve performance on) parent links creation in the graph.

Reply:
Yes good idea!